third_party/move/move-compiler-v2/src/pipeline/split_critical_edges_processor.rs

```diff
@@ -0,0 +1,200 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! This pass splits critical edges with empty blocks.
+//! A critical edge is an edge where the source node has multiple successors,
+//! and the target node has multiple predecessors.
+//!
+//! Side effects: clear existing annotations.
+//!
+//! Prerequisites: no call instructions have abort actions.
+//!
+//! Postconditions: no critical edges in the control flow graph.
+
+use move_model::{ast::TempIndex, model::FunctionEnv};
+use move_stackless_bytecode::{
+    function_target::FunctionData,
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    stackless_bytecode::{AttrId, Bytecode, Label},
+};
+use std::collections::{BTreeMap, BTreeSet};
+
+pub struct SplitCriticalEdgesProcessor {}
+
+impl FunctionTargetProcessor for SplitCriticalEdgesProcessor {
+    fn process(
+        &self,
+        _targets: &mut FunctionTargetsHolder,
+        fun_env: &FunctionEnv,
+        data: FunctionData,
+        _scc_opt: Option<&[FunctionEnv]>,
+    ) -> FunctionData {
+        if fun_env.is_native() {
+            return data;
+        }
+        let mut transformer = SplitCriticalEdgesTransformation::new(data);
+        transformer.transform();
+        transformer.data.annotations.clear();
+        transformer.data
+    }
+
+    fn name(&self) -> String {
+        "SplitCriticalEdgesProcessor".to_owned()
+    }
+}
+
+struct SplitCriticalEdgesTransformation {
+    data: FunctionData,
+    // labels used in the original code and in the generated code
+    labels: BTreeSet<Label>,
+    srcs_count: BTreeMap<Label, usize>,
+}
+
+impl SplitCriticalEdgesTransformation {
+    pub fn new(data: FunctionData) -> Self {
+        let labels = Bytecode::labels(&data.code);
+        let srcs_count = count_srcs(&data.code);
+        Self {
+            data,
+            labels,
+            srcs_count,
+        }
+    }
+
+    /// Runs the transformation, which breaks critical edges with empty blocks.
+    fn transform(&mut self) {
+        let bytecodes = std::mem::take(&mut self.data.code);
+        for bytecode in bytecodes {
+            self.transform_bytecode(bytecode)
+        }
+    }
+
+    /// Transforms a bytecode
+    fn transform_bytecode(&mut self, bytecode: Bytecode) {
+        match bytecode {
+            Bytecode::Branch(attr_id, l0, l1, t) => self.transform_branch(attr_id, l0, l1, t),
+            // Edge of a `Jump` can never be critical since the source node only have one out edge
+            _ => self.emit(bytecode),
+        }
+    }
+
+    /// Transforms a branch instruction by splitting the critical out edges
+    pub fn transform_branch(&mut self, attr_id: AttrId, l0: Label, l1: Label, t: TempIndex) {
+        match (
+            self.split_critical_edge(attr_id, l0),
```

Comment:
What if `l0 == l1` here?

Reply:
Nothing special would happen. If both branches are critical, then `if t goto l else goto l` is replaced by `if t goto l1 else goto l2` and new empty blocks `l1: goto l`, and `l2: goto l` are generated.

Reply:
Ok.